flang/lib/Lower/OpenMP.cpp

```diff
@@ -856,6 +893,96 @@ void Fortran::lower::genOpenMPConstruct(
       ompConstruct.u);
 }
 
+/// Generate the ThreadprivateOp strictly after addressing of variables or
+/// common block. Otherwise, instructions using addressed value between the
+/// the fir addressOfOp and omp threadprivateOp do not know if it has
+/// threadprivate flag or not.
+void Fortran::lower::genThreadprivateOp(
+    Fortran::lower::AbstractConverter &converter,
+    const Fortran::lower::pft::Variable &var) {
+  fir::FirOpBuilder &firOpBuilder = converter.getFirOpBuilder();
+  mlir::Location currentLocation = converter.getCurrentLocation();
+  MLIRContext *context = &converter.getMLIRContext();
+  mlir::Value symAddr;
+  mlir::StringAttr symName;
+
+  auto addThreadprivateSymAddr = [&](Fortran::lower::SymbolRef sym) {
+    symAddr = converter.getSymbolAddress(sym);
+  };
+
+  const Fortran::semantics::Symbol &sym = var.getSymbol();
+  if (const Fortran::semantics::Symbol *common =
+          Fortran::semantics::FindCommonBlockContaining(sym.GetUltimate())) {
+    addThreadprivateSymAddr(*common);
+    if (!symAddr.hasOneUse()) {
+      // Only generate ThreadprivateOp once for a common block following
+      // addressing the common block when instantiating the members of the
+      // common block to avoid redundancy.
+      return;
+    }
+    symName = mlir::StringAttr::get(context, converter.mangleName(*common));
+    // Generate ThreadprivateOp after addressing of common block.
+    auto insPt = firOpBuilder.saveInsertionPoint();
+    firOpBuilder.setInsertionPointAfterValue(symAddr);
+    firOpBuilder.create<mlir::omp::ThreadprivateOp>(currentLocation, symAddr,
+                                                    symName);
+    firOpBuilder.restoreInsertionPoint(insPt);
+    return;
+  }
+
+  if (!var.isGlobal()) {
+    // Non-global variable which can be in Threadprivate directive must be one
+    // variable in main program, and it has implicit SAVE attribute. Take it as
+    // with SAVE attribute, so to create GlobalOp for it to simplify the
+    // generation of pinned AllocaOp when entering OpenMP parallel region and
+    // the translation to LLVM IR.
+    mlir::Type ty = converter.genType(sym);
+    std::string globalName = converter.mangleName(sym);
+    mlir::StringAttr linkage = firOpBuilder.createInternalLinkage();
+    fir::GlobalOp global =
+        firOpBuilder.createGlobal(currentLocation, ty, globalName, linkage);
+    symName = global.getSymNameAttr();
+
+    auto createInit = [&](std::function<void(fir::FirOpBuilder &)> genInit) {
+      mlir::Region &region = global.getRegion();
+      region.push_back(new mlir::Block);
+      mlir::Block &block = region.back();
+      auto insertPt = firOpBuilder.saveInsertionPoint();
+      firOpBuilder.setInsertionPointToStart(&block);
+      genInit(firOpBuilder);
+      firOpBuilder.restoreInsertionPoint(insertPt);
+    };
+
+    // Create default initialization for non-character scalar.
+    if (Fortran::semantics::IsAllocatableOrPointer(sym)) {
```

Comment:
nit: Not entirely clear if the comment means ALLOCATABLE and POINTER attributes only apply to intrinsic types other than CHARACTER here or if the comment needs to be updated or?

Reply:
Yes, it only applies to non-character intrinsic types since the GlobalOp is emited for character and no need for this code.